### ISSUE TITLE
Allow UDF style array storage with store-like function

### ISSUE THREAD

**Issue by jakirkham** [dask/dask#4149]
It is repeatedly mentioned to us that `__setitem__` style `da.store` function doesn't work well for a majority of user cases as this means they need to open a file object and put in the graph. This then can fail for a variety of reasons like it is not safe to pickle, send across process boundaries, etc. Would be nice to have a form of `store` that takes a UDF that can handle all the file open/close logic as well as writing out the chunk.

cc @djhoese

**Comment by jakirkham** [comment]
Sure, so how I envisioned this (and admittedly was extrapolating from our previous conversations) is the user supplies a UDF where open/write/close happen within the UDF. We can imagine other arguments being passed along (e.g. filename, dataset name, etc.) to this `store`-like function. Thus a (typical) UDF might look like this.

```python
def my_udf_write(fn, dn, data):
    with h5py.File(fn, "a") as fh:
        fh[dn] = data
```

Please let me know if this fits with what you are thinking about or not.

**Comment by djhoese** [comment]
Ok I misunderstood UDF to mean a file object (and not user defined function). So with your example above, this would not allow writing blocks right? Or would slices also be provided?

The other thing I was hoping would be handled in whatever dask solutions comes out of this discussion is that the same solution would work with a threaded or distributed/multiprocess scheduler. So in a threaded scheduler, open the file once and reuse it for each block, closing automatically when all blocks are written. Perhaps an option for not closing it.

For a distributed scheduler, open it once per worker (per client/host maybe?), write the block, close it. As mentioned in #3255 I'm going to look at using xarray's CachingFileManager to help with this.

**Comment by jakirkham** [comment]
No worries. Thanks for the follow-up.

Yeah, not thinking file object. These are tricky things to have in the graph generally (some reasons for this noted in the OP). They work in the threaded case, but quickly become problematic outside of that case.

The UDF above would writing a block (or chunk); so, `data` would be that chunk.

Being scheduler friendly was one of the key motivators to the UDF design.

Currently, as proposed the UDF is open/closing when writing each chunk. Keeping the file open for all chunks is a bit tricky. My understanding is this is something xarray has wrestled with as well.

**Comment by djhoese** [comment]
> The UDF above would writing a block (or chunk); so, data would be that chunk.

The user would also need to know what slice of the overall array is being written, right?

> Currently, as proposed the UDF is open/closing when writing each chunk. Keeping the file open for all chunks is a bit tricky. My understanding is this is something xarray has wrestled with as well.

Yeah I think this is what the `CachingFileManager` is meant to help with.

**Comment by jakirkham** [comment]
We could add an [optional `block_info` like for `map_blocks`]( http://docs.dask.org/en/latest/array-api.html?highlight=block_info#dask.array.core.map_blocks ) if that would help.

I'm not sure that it opens the file once and keeps it open for the entire computation. Though I don't know too much about it.

**Comment by jakirkham** [comment]
> For a distributed scheduler, open it once per worker (per client/host maybe?), write the block, close it.

Should have mentioned explicitly that yes this is what the UDF is doing in the `with`.

**Comment by djhoese** [comment]
> We could add an optional block_info like for map_blocks if that would help.

Yeah that could work, but I also don't see a case where a chunked array wouldn't need to know this information. `map_blocks` having it optional makes sense since you are probably returning something with the same dimensions/shape as the input.

>> For a distributed scheduler, open it once per worker (per client/host maybe?), write the block, close it.

> Should have mentioned explicitly that yes this is what the UDF is doing in the with.

Yes, the point was that the "best" solution would be one that could do both the threaded and distributed scheduler handling without anything special on the users side.

**Comment by jakirkham** [comment]
Sure `block_info` could be required.

The UDF strategy seems like a reasonable start at least.

Are you familiar with [Zarr]( https://zarr.readthedocs.io/en/stable/ )? It works particularly well with Dask and can be [copied into other formats pretty easily (e.g. HDF5)]( https://zarr.readthedocs.io/en/stable/api/convenience.html#zarr.convenience.copy ). This is how we typically solve this problem at work.

**Comment by jakirkham** [comment]
Is this still an issue you are running into today, @djhoese?

**Comment by djhoese** [comment]
Yes and no. Dask hasn't added any interfaces like the ones we described above, has it?

I'd say my group has worked around it for the most part and our library (Satpy) does not have a lot of users running dask on multiprocess/cluster schedulers where this would be a bigger issue. Saving to zarr and then converting to another format is a workaround in my opinion and not a solution.

I also haven't kept up with what xarray has done as far as handling rasterio better or other similar situations.
